dex/src/main/java/io/crate/data/AsyncCompositeBatchIterator.java

```diff
@@ -117,9 +118,14 @@ public void close() {
 
             // Every "supplyAsync future" completes with the iterator::loadNextBatch() future, so in order to
             // access the loadNextBatch future, the asyncInvocation one should be completed and the result extracted
-            CompletableFuture<CompletableFuture<?>> supplyAsyncFuture =
-                CompletableFuture.supplyAsync(loadNextBatchFuture, executor);
-            asyncInvocationFutures.add(supplyAsyncFuture);
+            try {
+                CompletableFuture<CompletableFuture<?>> supplyAsyncFuture =
+                    CompletableFuture.supplyAsync(loadNextBatchFuture, executor);
+                asyncInvocationFutures.add(supplyAsyncFuture);
+            } catch (RejectedExecutionException e) {
+                // execute loadNextBatch synchronously if the executor can't currently take more tasks
+                iterator.loadNextBatch().toCompletableFuture().join();
```

Comment:
It's not necessary to block here, but rather the future should somehow be merged into `loadNextBatchFutures` later on.